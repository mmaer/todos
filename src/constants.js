export const LOCAL_STORAGE_KEY = "todos";

export const ENTER_KEY = 13;
export const ESCAPE_KEY = 27;

export const FILTER_ALL = "all";
export const FILTER_ACTIVE = "active";
export const FILTER_COMPLETED = "completed";

export const FILTERS = [FILTER_ALL, FILTER_ACTIVE, FILTER_COMPLETED];
