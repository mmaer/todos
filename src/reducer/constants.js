export const ADD_TODO = "ADD_TODO";
export const REMOVE_TODO = "REMOVE_TODO";
export const EDIT_TODO_TITLE = "EDIT_TODO_TITLE";

export const TOGGLE_COMPLETED_STATUS = "TOGGLE_COMPLETED_STATUS";
export const TOGGLE_EDITING_STATUS = "TOGGLE_EDITING_STATUS";

export const REMOVE_COMPLETED_TODOS = "REMOVE_COMPLETED_TODOS";
