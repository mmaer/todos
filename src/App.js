import React, { useEffect, useState } from "react";

import { useTodos } from "./reducer/actions";
import reducer from "./reducer";
import initialState from "./reducer/initialState";

import MainInput from "./components/MainInput";
import TodoList from "./components/TodoList";
import FilterBox from "./components/FilterBox";

import { filterActiveTodo, filterCompletedTodo } from "./utils";

import { LOCAL_STORAGE_KEY, FILTERS, FILTER_ALL } from "./constants";

import "./App.css";

const App = () => {
  const {
    state,
    addTodo,
    removeTodo,
    toggleCompletedStatus,
    toggleEditingStatus,
    editTodoTitle,
    removeCompletedTodos
  } = useTodos(
    JSON.parse(localStorage.getItem(LOCAL_STORAGE_KEY)) || initialState,
    reducer
  );

  useEffect(() => {
    localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(state));
  }, [state]);

  const [filter, setFilter] = useState(FILTER_ALL);

  useEffect(() => {
    const pathname = window.location.pathname.slice(1);
    if (FILTERS.includes(pathname)) {
      setFilter(pathname);
    } else {
      setFilter(FILTER_ALL);
    }
  }, []);

  const { todos } = state;

  return (
    <div>
      <header className="header">
        <h1>Todos</h1>
        <MainInput addTodo={addTodo} />
      </header>
      {!!todos.length && (
        <section className="main">
          <input className="toggle-all" type="checkbox" />
          <label htmlFor="toggle-all">Mark all as complete</label>
          <TodoList
            todos={todos}
            filter={filter}
            editTodoTitle={editTodoTitle}
            removeTodo={removeTodo}
            toggleCompletedStatus={toggleCompletedStatus}
            toggleEditingStatus={toggleEditingStatus}
          />
        </section>
      )}
      {!!todos.length && (
        <footer className="footer">
          <span className="todo-count">
            <strong>
              {todos.filter(filterActiveTodo).length}
              {` `}
            </strong>
            items left
          </span>
          <FilterBox filter={filter} setFilter={setFilter} />
          {!!todos.filter(filterCompletedTodo).length && (
            <button onClick={removeCompletedTodos} className="clear-completed">
              Clear completed
            </button>
          )}
        </footer>
      )}
    </div>
  );
};

export default App;
