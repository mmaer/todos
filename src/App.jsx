import React, { useEffect, useState } from "react";

import { useTodos } from "./reducer/actions";
import reducer from "./reducer";
import initialState from "./reducer/initialState";

import AddTodo from "./components/AddTodo";
import TodoList from "./components/TodoList";
import Footer from "./components/Footer";
import ToggleCompletedStatuses from "./components/ToggleCompletedStatuses";

import { LOCAL_STORAGE_KEY, FILTERS, FILTER_ALL } from "./constants";

import "./App.scss";

const App = () => {
  const {
    state,
    addTodo,
    removeTodo,
    toggleCompletedStatus,
    toggleEditingStatus,
    editTodoTitle,
    removeCompletedTodos,
    toggleCompletedStatuses
  } = useTodos(
    JSON.parse(localStorage.getItem(LOCAL_STORAGE_KEY)) || initialState,
    reducer
  );

  useEffect(() => {
    localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(state));
  }, [state]);

  const [filter, setFilter] = useState(FILTER_ALL);

  useEffect(() => {
    const pathname = window.location.pathname.slice(1);
    if (FILTERS.includes(pathname)) {
      setFilter(pathname);
    } else {
      setFilter(FILTER_ALL);
    }
  }, []);

  const { todos } = state;

  return (
    <div className="app">
      <header>
        <h1 className="app__title">Todos</h1>
        <AddTodo addTodo={addTodo} />
      </header>
      {!!todos.length && (
        <>
          <section className="app__main">
            <ToggleCompletedStatuses
              todos={todos}
              toggleCompletedStatuses={toggleCompletedStatuses}
            />
            <TodoList
              todos={todos}
              filter={filter}
              editTodoTitle={editTodoTitle}
              removeTodo={removeTodo}
              toggleCompletedStatus={toggleCompletedStatus}
              toggleEditingStatus={toggleEditingStatus}
            />
          </section>
          <Footer
            todos={todos}
            filter={filter}
            setFilter={setFilter}
            removeCompletedTodos={removeCompletedTodos}
          />
        </>
      )}
    </div>
  );
};

export default App;
