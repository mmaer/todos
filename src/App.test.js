import React from "react";
import { MemoryRouter } from "react-router-dom";
import { mount } from "enzyme";

import { ENTER_KEY } from "./constants";

import App from "./App";

describe("App component", () => {
  it("Should match snapshot", () => {
    const wrapper = mount(<App />);
    expect(wrapper).toMatchSnapshot();
  });

  it("Should display todo list when todo is added", () => {
    const wrapper = mount(
      <MemoryRouter>
        <App />
      </MemoryRouter>
    );
    const addTodoInput = wrapper.find(".add-todo");

    addTodoInput.simulate("change", { target: { value: "new-todo" } });
    addTodoInput.simulate("keyPress", { persist: () => {}, which: ENTER_KEY });

    expect(wrapper.find(".app__main").length).toBe(1);
  });
});
