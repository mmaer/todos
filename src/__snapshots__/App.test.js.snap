// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component Should match snapshot 1`] = `
<App>
  <div
    className="app"
  >
    <header>
      <h1
        className="app__title"
      >
        Todos
      </h1>
      <Memo()
        addTodo={[Function]}
      >
        <input
          className="add-todo"
          onChange={[Function]}
          onKeyPress={[Function]}
          placeholder="What needs to be done?"
          value=""
        />
      </Memo()>
    </header>
  </div>
</App>
`;
